### PR TITLE
Safer console

### PR DESCRIPTION
**Pull request by Lokathor**
TODO:
* None?

### CODE REVIEW

**Comment by Lokathor** [comment]
Okay, looks like I've now safe-ified all functions that don't touch a `HANDLE`.

File: src/wincon.rs [hunk 1]
```diff
@@ -65,16 +65,19 @@ pub fn add_console_alias(
   }
 }
 
-/// Allocates a new console for the process.
+/// Attatches your process to the console of the process you specify.
```
Comment (MaulingMonkey) [inline src/wincon.rs]: Extra T in "Attaches"

File: src/wincon.rs [hunk 2]
```diff
@@ -272,31 +304,41 @@ pub unsafe fn get_console_font_size(
...
+///
+/// Naturally, this fails if this process has no console.
+///
 /// See [`GetConsoleHistoryInfo`](https://docs.microsoft.com/en-us/windows/console/GetConsoleHistoryInfo)
 #[inline]
-pub unsafe fn get_console_history_info() -> Result<CONSOLE_HISTORY_INFO, ErrorCode> {
+pub fn get_console_history_info() -> Result<CONSOLE_HISTORY_INFO, ErrorCode> {
   let mut out = CONSOLE_HISTORY_INFO::default();
-  if GetConsoleHistoryInfo(&mut out) != 0 {
+  out.cbSize = size_of::<CONSOLE_HISTORY_INFO>() as u32;
```
Comment (MaulingMonkey) [inline src/wincon.rs]: (aside: I tend to prefer `size_of_val(&out)` in case types change but this should be valid)

Comment (Lokathor) [reply src/wincon.rs]: That would have no difference.

The only time that there is a difference between `size_of` and `size_of_val` is when T is a dynamically sized type.

Comment (MaulingMonkey) [reply src/wincon.rs]: I'm talking about the scenario where CONSOLE_HISTORY_INFO gets replaced with CONSOLE_HISTORY_INFO_EX or something.  I've seen the occasional bug where during such a refactor, only one of the types gets changed, causing out.cbSize != sizeof(out), causing sorrow.  It's rare, but it happens, and I'm paranoid ;)

File: src/wincon.rs [hunk 3]
```diff
@@ -82,6 +83,31 @@ impl ConsoleScreenBuffer {
...
+  /// you. Any other bit of code can still get the stdout handle and then do
+  /// output (eg: `println!`). You could even call this more than once and have
+  /// more than one `ConsoleScreenBuffer` all pointing at stdout.
+  ///
+  /// The `ConsoleScreenBuffer` is wrapped in `ManuallyDrop` so that your stdout
+  /// handle doesn't get closed when you drop it. You do not need to manually
+  /// drop anything to use this method. Just let go of the value when you're
+  /// done with it and everything will be fine.
+  #[cfg(feature = "processenv")]
+  pub fn from_stdout() -> Result<ManuallyDrop<Self>, ErrorCode> {
```
Comment (MaulingMonkey) [inline src/wincon.rs]: I'd be tempted to DuplicateHandle ([msdn claims](https://docs.microsoft.com/en-us/windows/console/console-handles) this is fine) to eliminate the need to leak ManuallyDrop into the public interface.

Comment (Lokathor) [reply src/wincon.rs]: I'll look at this next time I sit down with the lib.

File: src/consoleapi.rs [hunk 4]
```diff
@@ -12,16 +12,19 @@ use winapi::um::consoleapi::{
   SetConsoleMode, WriteConsoleW,
 };
 
-/// Allocates a new console for the process.
+/// Allocates a new console for the process and attatches to it.
 ///
-/// A process can only have one console at a time. A console process has a
-/// console by default, but a GUI process does not.
+/// A process can only be attatched to one console at a time. A "console"
```
Comment (memoryruins) [inline src/consoleapi.rs]:
```suggestion
/// A process can only be attached to one console at a time. A "console"
```

File: src/consoleapi.rs [hunk 5]
```diff
@@ -12,16 +12,19 @@ use winapi::um::consoleapi::{
...
+/// process starts with one, but a "GUI" process does not.
+///
+/// If you try to allocate a console while your process is already attatched to
```
Comment (memoryruins) [inline src/consoleapi.rs]:
```suggestion
/// If you try to allocate a console while your process is already attached to
```

**Pull request merged**